go-controller/pkg/util/libovsdb.go

```diff
@@ -13,13 +13,177 @@ import (
 	"github.com/cenkalti/backoff/v4"
 	"github.com/ovn-org/libovsdb/client"
 	"github.com/ovn-org/libovsdb/model"
+	"github.com/ovn-org/libovsdb/ovsdb"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/sbdb"
 	"gopkg.in/fsnotify/fsnotify.v1"
 	"k8s.io/klog/v2"
 )
 
+type ModelClient struct {
+	nbclient client.Client
+}
+
+func NewModelClient(nbclient client.Client) ModelClient {
+	return ModelClient{nbclient: nbclient}
+}
+
+// OperationModel is a struct which uses reflection to determine and
+// perform idempotent operations against NB DB
+type OperationModel struct {
+	// Model specifies the model to be created/updated/deleted
+	Model interface{}
+	// ModelPredicate specifies the predicate at which the lookup for the model
+	// is made. This is what defines if the object exists or not. Required
+	// field.
+	ModelPredicate interface{}
+	// ModelMutations specifies the mutations to be performed on the model if an
+	// item exists. This is an optional field, and not specifying it means that
+	// a create/update operation is to be performed.
+	ModelMutations []model.Mutation
+	// ModelUpdates specifies the model fields to be updated if an item exists.
+	// This is an optional field, and not specifying it means that a
+	// create/mutation operation is to be performed. If specified: ModelCreate
+	// needs to be specified too, as that is used to demark which values each
+	// field to be update will have.
+	ModelUpdates []interface{}
+	// ReferentialModel specifies an eventual model which has a reference to the
+	// model being created. It can only be used in conjunction with a model
+	// create. This field is optional and only needed for "sub models" which
+	// have references to other "parent models", ex: logicalRouterPolicies ->
+	// LogicalRouter. This field also only needs to specify the type of the
+	// model (for ease of introspection), the actual object is looked up using
+	// ReferentialModelPredicate
+	ReferentialModel interface{}
```

Comment:
Yeah, it could I guess. So essentially you're saying that `ReferentialModel` would be a pointer to another `OperationModel`?

Reply:
Perhaps we just need an array of OperationalModels  and remove ReferentialModel.

So from
```
	opModel = util.OperationModel{
		Model:            &logicalSwitchPort,
		ModelPredicate:   func(lsp *nbdb.LogicalSwitchPort) bool { return lsp.Name == gwSwitchPort },
		ReferentialModel: &logicalSwitch,
		ReferentialModelMutation: model.Mutation{
			Field:   &logicalSwitch.Ports,
			Mutator: ovsdb.MutateOperationInsert,
			Value:   []string{gwSwitchPort},
		},
		ReferentialModelPredicate: func(ls *nbdb.LogicalSwitch) bool { return ls.Name == types.OVNJoinSwitch },
	}
```
to
```
	opModel = []util.OperationModel{
            {
		Model:            &logicalSwitchPort,
		ModelPredicate:   func(lsp *nbdb.LogicalSwitchPort) bool { return lsp.Name == gwSwitchPort },
             },
            {
		Model: &logicalSwitch,
		ModelMutation: model.Mutation{
			Field:   &logicalSwitch.Ports,
			Mutator: ovsdb.MutateOperationInsert,
			Value:   []string{gwSwitchPort},
		},
		ModelPredicate: func(ls *nbdb.LogicalSwitch) bool { return ls.Name == types.OVNJoinSwitch },
            },
	}
```

ModelPredicate could be optional, the Model could be used for the search if indexes are provided.
Also could be nice to infer ExistingResult.

Reply:
Oh I see, indeed that would make the model simpler. I will change that now. 

> Also could be nice to infer ExistingResult

Again, I agree. I tried for a couple of hours but however gave up fighting `reflect` it on this front. Maybe you can update that part, if you feel more comfortable using that package than I do?

Reply:
Actually, I realized one corner case that is not covered by this proposed model and renders it less efficient, @jcaamano. This applies to updates of existing objects, so say for example that an object in its current form already exists in the DB, but one field needs to be updated. 

Previously one could do:

```go
        logicalRouter := nbdb.LogicalRouter{}
	logicalRouterPolicy := nbdb.LogicalRouterPolicy{
		Nexthops: []string{nexthops},
		Priority:  priority,
		Match:    match,
		Action:   nbdb.LogicalRouterPolicyActionReroute,
		UUID:     namedUUID,
	}

	opModels := OperationModel{
		{
			Model: &logicalRouterPolicy,
			ModelPredicate: func(lrp *nbdb.LogicalRouterPolicy) bool {
				return lrp.Priority == priority
			},
			OnModelUpdates: []interface{}{
				&logicalRouterPolicy.Action,
			},
                        ReferentialModel:          &logicalRouter,
			ReferentialModelPredicate: func(lr *nbdb.LogicalRouter) bool { return lr.Name == types.OVNClusterRouter },
			OnReferentialModelMutations: []model.Mutation{
				{
					Field:   &logicalRouter.Policies,
					Mutator: ovsdb.MutateOperationInsert,
					Value:   []string{namedUUID},
				},
			},
			ExistingResult: &[]nbdb.LogicalRouterPolicy{},
		},
        if _, err := modelClient.CreateOrUpdate(opModels...); err != nil {
		return fmt.Errorf("unable to create policy based routes, err: %v", err)
	}
```

That would have the model created if it does not exist, and update the `.Action` field if it did exist, while also not adding the UUID to the "referential model's" `.Policies`

Currently one can write:

```go
	logicalRouter := nbdb.LogicalRouter{}
	logicalRouterPolicy := nbdb.LogicalRouterPolicy{
		Nexthops: []string{nexthops},
		Priority: priority,
		Match:    match,
		Action:   nbdb.LogicalRouterPolicyActionReroute,
		UUID:     namedUUID,
	}

	opModels := []OperationModel{
		{
			Model: &logicalRouterPolicy,
			ModelPredicate: func(lrp *nbdb.LogicalRouterPolicy) bool {
				return lrp.Priority == priority
			},
			OnModelUpdates: []interface{}{
				&logicalRouterPolicy.Action,
			},
			ExistingResult: &[]nbdb.LogicalRouterPolicy{},
		},
		{
			Model:          &logicalRouter,
			ModelPredicate: func(lr *nbdb.LogicalRouter) bool { return lr.Name == types.OVNClusterRouter },
			OnModelMutations: []model.Mutation{
				{
					Field:   &logicalRouter.Policies,
					Mutator: ovsdb.MutateOperationInsert,
					Value:   []string{namedUUID},
				},
			},
			ExistingResult: &[]nbdb.LogicalRouter{},
		},
	}
	if _, err := modelClient.CreateOrUpdate(opModels...); err != nil {
		return fmt.Errorf("unable to create policy based routes, err: %v", err)
	}
```

but the second `OperationModel` will break as it references a named UUID which does not exists. I could treat all operations model in order and break whenever there's a model which requires an update (specifies `OnModelUpdate`) and thus never treat the "referential model" which would bresk, but I am not sure that's better as it makes the behavior implicit.

Reply:
I see what you say. If we want to be implicit and generic we need to be smarter about this references. So instead of `Value:   []string{namedUUID}` we could have `WithReferences: []Model{&logicalRouterPolicy}` that we would translate internally to `Value:   []string{logicalRouterPolicy.uuid}}` .  `logicalRouterPolicy.uuid` would be the original named uuid if it didn't exist or would be updated with the actual uuid if it did exist.

Reply:
> So instead of `Value: []string{namedUUID}` we could have `WithReferences: []Model{&logicalRouterPolicy}` that we would translate internally to `Value: []string{logicalRouterPolicy.uuid}}` .

That specific change would require a change to libovsdb. We could however maybe add a wrapper around `model.Mutation` to a "local type" and convert that following the procedure you mentioned, but I am not sure I would be a fan of that and would probably prefer keeping the old model.

Reply:
Yes, we would need our own `ModelMutation` or some other field in `OperationModel`.

And I don't see a way around it. Old model is not solving everything, just a specific case. I need to `CreateOrUpdate` acls, ports and a port group with references to both. I will need to deal with it. Unless we accept to do multiple operations for things that could be done in less but I did want to avoid it if possible.

Reply:
OK, so in the mean time I did this....but I think I am loosing perspective a bit :smile: 

```go
	logicalRouter := nbdb.LogicalRouter{}
	logicalRouterPolicy := nbdb.LogicalRouterPolicy{
		Nexthops: []string{nexthops},
		Priority: intPriority,
		Match:    match,
		Action:   nbdb.LogicalRouterPolicyActionReroute,
		UUID:     namedUUID,
	}

	result := []nbdb.LogicalRouterPolicy{}

	opModels := []util.OperationModel{
		{
			Model: &logicalRouterPolicy,
			ModelPredicate: func(lrp *nbdb.LogicalRouterPolicy) bool {
				return strings.Contains(lrp.Match, fmt.Sprintf("%s\"", nodeName)) && lrp.Priority == intPriority
			},
			OnModelUpdates: []interface{}{
				&logicalRouterPolicy.Match,
				&logicalRouterPolicy.Nexthops,
			},
			ExistingResult: &result,
		},
		{
			Model: &logicalRouter,
			ModelPredicate: func(lr *nbdb.LogicalRouter) bool {
				return lr.Name == types.OVNClusterRouter
			},
			OnModelMutations: func() []model.Mutation {
				uuid := namedUUID
				if len(result) > 0 {
					uuid = result[0].UUID
				}
				return []model.Mutation{
					{
						Field:   &logicalRouter.Policies,
						Mutator: ovsdb.MutateOperationInsert,
						Value:   []string{uuid},
					},
				}
			},
			ExistingResult: &[]nbdb.LogicalRouter{},
		},
	}
	if _, err := modelClient.CreateOrUpdate(opModels...); err != nil {
		return fmt.Errorf("unable to create policy based routes, err: %v", err)
	}
```

So, essentially making `OnModelMutations` a callback which allows us to evaluate UUID references "after the fact". I am not convinced this is good API, but it does solve the use cases that I at least have and lead to less code duplication.

Reply:
Taking a step back. The problem we want to solve is to batch `CreateOrUpdate` using named UUIDs.

Let's say that you start with a `LogicalRouter` and `LogicalRouterPolicy` already using named UUIDs: 

```
logicalRouter := nbdb.LogicalRouter{
  Name: OVNClusterRouter,
  Policies: namedUUID,
}
logicalRouterPolicy := nbdb.LogicalRouterPolicy{
  Nexthops: []string{nexthops},
  Priority: intPriority,
  Match:    match,
  Action:   nbdb.LogicalRouterPolicyActionReroute,
  UUID:    namedUUID,
}
```
As you go along checking for existence, you can save a map of `namedUUID -> uuid`

Then you could have
```
{
	Model: &logicalRouter,
        OnModelReferences: []interface{}{
		&logicalRouter.Policies,
	},
},
```

Then you would know you would have to go into `logicalRouter.Policies` and replace any named uuid there with what you have in the map. Then if `logicalRouter` does not exist is a straight forward create. If it does exist, we can internally infer the mutation pretty much assuming an insert Mutation.

Reply:
> Taking a step back. The problem we want to solve is to batch `CreateOrUpdate` using named UUIDs.

Well, not only for named UUIDs. I would like to have a mechanism whereby we can create an object if it does not exist (and generate the reference to any other object - should it be required) or update/mutate the object if it does exists and if the client specifies that it wants to do that.    

I am still not sure that `OnModelReferences` manages to cover this use case. In the example I gave above: https://github.com/ovn-org/ovn-kubernetes/pull/2344#discussion_r673925406 if we use `OnModelReferences` and do:

```go
	logicalRouter := nbdb.LogicalRouter{}
	logicalRouterPolicy := nbdb.LogicalRouterPolicy{
		Nexthops: []string{nexthops},
		Priority: intPriority,
		Match:    match,
		Action:   nbdb.LogicalRouterPolicyActionReroute,
		UUID:     namedUUID,
	}

	result := []nbdb.LogicalRouterPolicy{}

	opModels := []util.OperationModel{
		{
			Model: &logicalRouterPolicy,
			ModelPredicate: func(lrp *nbdb.LogicalRouterPolicy) bool {
				return strings.Contains(lrp.Match, fmt.Sprintf("%s\"", nodeName)) && lrp.Priority == intPriority
			},
			OnModelUpdates: []interface{}{
				&logicalRouterPolicy.Match,
				&logicalRouterPolicy.Nexthops,
			},
			ExistingResult: &result,
		},
		{
			Model: &logicalRouter,
			ModelPredicate: func(lr *nbdb.LogicalRouter) bool {
				return lr.Name == types.OVNClusterRouter
			},
			OnModelReferences: []interface{} {
				&logicalRouterPolicy,
			},
			ExistingResult: &[]nbdb.LogicalRouter{},
		},
	}
	if _, err := modelClient.CreateOrUpdate(opModels...); err != nil {
		return fmt.Errorf("unable to create policy based routes, err: %v", err)
	}
``` 

and have `OnModelReferences` be converted to `model.Mutations` in `CreateOrUpdate`: I don't see how this handles the update case, specifically: if the object exists we don't want to re-reference the logical router policy by the logical router, or at least reference the existing UUID value (so that it doesn't error in OVS DB). By specifying this we would be using the `namedUUID` value that has been specified on the  `logicalRouterPolicy` object in case it needs to be created.  

Using a map to track all UUIDs, and specifically named UUID -> UUID seems more complex than the callback function I currently have. We would have to deal with thread safety for this map and properly syncing it. Moreover, the callback could be used to do more complicated evaluations going further, i.e: not only mutate when the object is being created, but also based on other conditions. 

I think that the one thing that can be simplified now from this model is not needing to specify `ExisitingResults` if it is not provided, and infer and create it based on the `Model`

Reply:
> Well, not only for named UUIDs. I would like to have a mechanism whereby we can create an object if it does not exist (and generate the reference to any other object - should it be required) or update/mutate the object if it does exists and if the client specifies that it wants to do that.

Yes, I over-simplified. That would also be the intention.

> 
> I am still not sure that `OnModelReferences` manages to cover this use case. In the example I gave above: [#2344 (comment)](https://github.com/ovn-org/ovn-kubernetes/pull/2344#discussion_r673925406) if we use `OnModelReferences` and do:
> 
> ```go
> 	logicalRouter := nbdb.LogicalRouter{}
> 	logicalRouterPolicy := nbdb.LogicalRouterPolicy{
> 		Nexthops: []string{nexthops},
> 		Priority: intPriority,
> 		Match:    match,
> 		Action:   nbdb.LogicalRouterPolicyActionReroute,
> 		UUID:     namedUUID,
> 	}
> 
> 	result := []nbdb.LogicalRouterPolicy{}
> 
> 	opModels := []util.OperationModel{
> 		{
> 			Model: &logicalRouterPolicy,
> 			ModelPredicate: func(lrp *nbdb.LogicalRouterPolicy) bool {
> 				return strings.Contains(lrp.Match, fmt.Sprintf("%s\"", nodeName)) && lrp.Priority == intPriority
> 			},
> 			OnModelUpdates: []interface{}{
> 				&logicalRouterPolicy.Match,
> 				&logicalRouterPolicy.Nexthops,
> 			},
> 			ExistingResult: &result,
> 		},
> 		{
> 			Model: &logicalRouter,
> 			ModelPredicate: func(lr *nbdb.LogicalRouter) bool {
> 				return lr.Name == types.OVNClusterRouter
> 			},
> 			OnModelReferences: []interface{} {
> 				&logicalRouterPolicy,
> 			},
> 			ExistingResult: &[]nbdb.LogicalRouter{},
> 		},
> 	}
> 	if _, err := modelClient.CreateOrUpdate(opModels...); err != nil {
> 		return fmt.Errorf("unable to create policy based routes, err: %v", err)
> 	}
> ```
> 
> and have `OnModelReferences` be converted to `model.Mutations` in `CreateOrUpdate`: I don't see how this handles the update case, specifically: if the object exists we don't want to re-reference the logical router policy by the logical router, or at least reference the existing UUID value (so that it doesn't error in OVS DB). By specifying this we would be using the `namedUUID` value that has been specified on the `logicalRouterPolicy` object in case it needs to be created.

`OnModelReferences` would not be a substitute for `OnModelUpdates` or `OnModelMutations`. It is just an indication of properties that would need to be inspected for replacement of named UUIDs for actual UUIDs, and doing so because it is a general issue that could be solved in a generic way. How we handle updates vs mutations would be a different concern, but given that the main operation is called `CreateOrUpdate` perhaps we could infer it as well, or maybe not but solve it in a different way. I don't see why indicating an `update` vs indicating an `insert` would make an orthogonal difference to an API that is useful to us. So maybe `OnModelUpdates` and `OnModelInserts` and `OnModelDeletes` with references to struct members for each might provide a generic way to express what needs to happen?

> Using a map to track all UUIDs, and specifically named UUID -> UUID seems more complex than the callback function I currently have. We would have to deal with thread safety for this map and properly syncing it. Moreover, the callback could be used to do more complicated evaluations going further, i.e: not only mutate when the object is being created, but also based on other conditions.

Named UUIDs context is a single transaction so I don't think there is any thread safety concerns?

I see how a callback makes up for increased flexibility. But now I am confused if we are trying to provide flexibility or to simplify something that would otherwise be too repetitive.

> 
> I think that the one thing that can be simplified now from this model is not needing to specify `ExisitingResults` if it is not provided, and infer and create it based on the `Model`
 
Agreed.

Reply:
I guess that I have to say to go ahead with what is working for you now in any case. May main gripe with the callback for specifically solving named uuids is that it requires a locality that would otherwise not be required and I think it is relevant on a situation where we might batch operations from different contexts but that we can still correlate with our own internal ids.